### PR TITLE
[Feature store] SDK classes + API + local ingestion and retrieval

### PR DESCRIPTION
**Pull request by yaronha**
* add feature store model/sdk, ingestion and retrival

### CODE REVIEW

File: tests/feature-store/test_feature_store.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+from mlrun.featurestore.targets import CSVTarget
+from mlrun.utils import logger
+import mlrun.featurestore as fs
+from mlrun.config import config as mlconf
+from mlrun.featurestore import FeatureSet, Entity, run_ingestion_task
+from mlrun.featurestore.model.datatypes import ValueType
+from mlrun.featurestore.model.validators import MinMaxValidator
+
+
+this_dir = f"{tests_root_directory}/feature-store/"
```
Comment (Hedingber) [inline tests/feature-store/test_feature_store.py:23]:
```suggestion
local_dir = f"{tests_root_directory}/feature-store/"
```

File: tests/feature-store/test_infer.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    {"name": "movements_is_error", "value_type": "bool"},
+    {"name": "turn_count", "value_type": "float"},
+    {"name": "turn_count_is_error", "value_type": "bool"},
+    {"name": "is_in_bed", "value_type": "int"},
+    {"name": "is_in_bed_is_error", "value_type": "bool"},
+    {"name": "timestamp", "value_type": "str"},
+]
+
+
+def test_infer():
```
Comment (Hedingber) [inline tests/feature-store/test_infer.py:28]:
```suggestion
def test_infer_from_df():
```

**Review by Hedingber — CHANGES_REQUESTED** [review]
* feature set/feature vector/feature store are all two words, not one, so it should be (file names and variable names) - `feature_store`, `feature_set`, `feature_vector` please fix everywhere
* In a lot of examples (in docstrings) not all variables are defined, please do a pass and make sure everything is defined
* Do a pass and change to use mlrun.errors everywhere

File: mlrun/featurestore/__init__.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    "get_offline_features",
+    "get_online_feature_service",
+    "ingest",
+    "infer_metadata",
+    "run_ingestion_task",
+    "Feature",
+    "Entity",
+    "FeatureSet",
+    "FeatureVector",
+]
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/__init__.py:25]: i want to only export things that are well documented and useful for users (vs internal)

File: mlrun/featurestore/__init__.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+from .model.validators import MinMaxValidator  # noqa
+from .api import (
+    get_offline_features,
+    get_online_feature_service,
+    ingest,
+    infer_metadata,
+    run_ingestion_task,
+)  # noqa
+from .targets import TargetTypes  # noqa
+from .infer import InferOptions  # noqa
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/__init__.py:40]: done

File: mlrun/featurestore/api.py [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+
+
+def get_offline_features(
+    features: Union[str, List[str], FeatureVector],
+    entity_rows=None,
+    entity_timestamp_column: str = None,
+    name: str = None,
+    watch: bool = True,
+    store_target: DataTargetSpec = None,
+):
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/api.py:64]: done

File: mlrun/featurestore/api.py [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+
+def _features_to_vector(features, name):
+    if isinstance(features, str):
+        vector = get_feature_vector_by_uri(features)
+    elif isinstance(features, list):
+        vector = FeatureVector(features=features)
+    elif isinstance(features, FeatureVector):
+        vector = features
+    else:
+        raise ValueError("illegal features value/type")
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/api.py:50]: done (for all)

File: mlrun/featurestore/api.py [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+
+def get_offline_features(
+    features: Union[str, List[str], FeatureVector],
+    entity_rows=None,
+    entity_timestamp_column: str = None,
+    name: str = None,
+    watch: bool = True,
+    store_target: DataTargetSpec = None,
+):
+    """retrieve offline feature vector
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/api.py:65]: its a common feature store convention/naming (like other frameworks), we dont return the vectors, rather a response w features (and vector object is stored/linked in the resp object)

File: mlrun/featurestore/api.py [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+            "stock-quotes.ask as mycol",
+            "stocks.*",
+        ]
+
+        resp = fs.get_offline_features(
+            features, entity_rows=trades, entity_timestamp_column="time"
+        )
+        print(resp.to_dataframe())
+        print(resp.vector.get_stats_table())
+        resp.to_parquet("./out.parquet")
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/api.py:81]: afaik inline examples are not full programs and dont need to include the data or imports

[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/api.py:81]: i will remove the `fs.` (the function is the one we doc)

File: mlrun/featurestore/api.py [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+        print(resp)
+        resp = svc.get([{"ticker": "AAPL"}])
+        print(resp)
+
+    :param features:     list of features or feature vector uri or FeatureVector object
+    :param name:         name to use for the generated feature vector
+    :param function:     optional, mlrun FunctionReference object, serverless function template
+    """
+    vector = _features_to_vector(features, name)
+    controller = init_feature_vector_graph(vector)
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/api.py:114]: it initializes the graph, and returns with the graph controller (it doesnt init the controller)

File: mlrun/featurestore/common.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+# limitations under the License.
+
+import mlrun
+from mlrun.utils import parse_versioned_object_uri
+from ..config import config
+
+feature_separator = "."
+
+
+def parse_features(vector):
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/common.py:22]: we can change, was there originally dont remember why i changed

File: mlrun/featurestore/common.py [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+
+
+def _parse_feature_string(feature):
+    """parse feature string into feature set name, feature name, alias"""
+    # expected format: <feature-set>.<name|*>[ as alias]
+    if feature_separator not in feature:
+        raise ValueError(
+            f"feature {feature} must be in the form feature-set.feature[ as alias]"
+        )
+    splitted = feature.split(feature_separator)
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/common.py:86]: done

File: mlrun/featurestore/infer.py [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+    current_entities = list(featureset_spec.entities.keys())
+    entity_columns = entity_columns or []
+
+    def upsert_feature(name, value_type):
+        if name in featureset_spec.features:
+            featureset_spec.features[name].value_type = value_type
+        else:
+            featureset_spec.features[name] = Feature(name=column, value_type=value_type)
+
+    for column, s in df.items():
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/infer.py:75]: `series`, updates

File: mlrun/featurestore/ingestion.py [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+
+    # init targets (and table)
+    targets = targets or []
+    _add_data_states(
+        graph, cache, featureset, targets=targets, source=source, return_df=return_df,
+    )
+
+    server = create_graph_server(graph=graph, parameters={})
+    server.init(None, namespace, cache)
+    return graph._controller
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/ingestion.py:39]: i dont use it, but i can return graph if it makes u happy :)

File: mlrun/featurestore/ingestion.py [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+    graph.set_flow_source(source)
+
+
+def deploy_ingestion_function(
+    name, source, featureset, parameters, function_ref=None, local=False, watch=True
+):
+    name = name or f"{featureset.metadata.name}_ingest"
+    function_ref = function_ref or featureset.spec.function.copy()
+    if not function_ref.to_dict():
+        runtime = RuntimeKinds.serving if source.online else RuntimeKinds.job
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/ingestion.py:90]: done

File: mlrun/featurestore/model/base.py [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+    parquet = "parquet"
+    nosql = "nosql"
+    tsdb = "tsdb"
+    stream = "stream"
+    dataframe = "dataframe"
+
+
+class ResourceKinds:
+    FeatureSet = "FeatureSet"
+    FeatureVector = "FeatureVector"
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/model/base.py:40]: removed

File: mlrun/featurestore/api.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/api.py:65]: fixed doc/text

File: mlrun/featurestore/api.py [hunk 16]
```diff
@@ -0,0 +1,274 @@
...
+    vector = _features_to_vector(features, name)
+    entity_timestamp_column = entity_timestamp_column or vector.spec.timestamp_field
+    merger = LocalFeatureMerger(vector)
+    return merger.start(entity_rows, entity_timestamp_column, store_target)
+
+
+def get_online_feature_service(
+    features: Union[str, List[str], FeatureVector], name: str = None, function=None
+):
+    """initialize and return the feature vector online client
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/api.py:99]: fixed text

File: mlrun/featurestore/api.py [hunk 17]
```diff
@@ -0,0 +1,274 @@
...
+    example::
+
+        svc = fs.get_online_feature_service(vector_uri)
+        resp = svc.get([{"ticker": "GOOG"}, {"ticker": "MSFT"}])
+        print(resp)
+        resp = svc.get([{"ticker": "AAPL"}])
+        print(resp)
+
+    :param features:     list of features or feature vector uri or FeatureVector object
+    :param name:         name to use for the generated feature vector
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/api.py:110]: removed name

File: mlrun/featurestore/api.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/api.py:114]: done

File: mlrun/featurestore/api.py [hunk 18]
```diff
@@ -0,0 +1,274 @@
...
+    namespace=None,
+    return_df: bool = True,
+    infer_options: InferOptions = InferOptions.Null,
+):
+    """Read local DataFrame, file, or URL into the feature store
+
+    example::
+
+        stocks_set = fs.FeatureSet("stocks", entities=[Entity("ticker")])
+        df = ingest(stocks_set, stocks, infer_options=fs.InferOptions.default())
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/api.py:132]: done (added read_csv)

File: mlrun/featurestore/common.py [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+            )
+        feature = feature_set_object[name]
+        processed_features[alias or name] = (feature_set_object, feature)
+        featureset_name = feature_set_object.metadata.name
+        if featureset_name in feature_set_fields.keys():
+            value = feature_set_fields[featureset_name]
+            value.append((name, alias))
+            feature_set_fields[featureset_name] = value
+        else:
+            feature_set_fields[featureset_name] = [(name, alias)]
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/common.py:42]: done

File: mlrun/featurestore/common.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/common.py:22]: done

File: mlrun/featurestore/ingestion.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/ingestion.py:39]: done

File: mlrun/featurestore/model/base.py [hunk 20]
```diff
@@ -0,0 +1,270 @@
...
+
+from mlrun.model import ModelObj
+from .datatypes import ValueType
+from .validators import validator_types
+
+
+class FeatureStoreError(Exception):
+    """error in feature store data or configuration"""
+
+    pass
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/model/base.py:26]: done

File: mlrun/featurestore/model/base.py [hunk 21]
```diff
@@ -0,0 +1,270 @@
...
+        self.uri = uri
+        self.sources = sources or {}
+
+
+class SourceTypes:
+    csv = "csv"
+    dataframe = "dataframe"
+
+
+class DataTargetSpec(ModelObj):
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/model/base.py:156]: changed to Base

File: mlrun/featurestore/model/base.py [hunk 22]
```diff
@@ -0,0 +1,270 @@
...
+        self.windows = windows or []
+        self.period = period
+
+
+class CommonMetadata(ModelObj):
+    def __init__(
+        self,
+        name: str = None,
+        tag: str = None,
+        hash: str = None,
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/model/base.py:258]: changed

File: mlrun/featurestore/model/featureset.py [hunk 23]
```diff
@@ -0,0 +1,361 @@
...
+
+    @property
+    def status(self) -> FeatureSetStatus:
+        return self._status
+
+    @status.setter
+    def status(self, status):
+        self._status = self._verify_dict(status, "status", FeatureSetStatus)
+
+    def uri(self):
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/model/featureset.py:215]: property

File: mlrun/featurestore/model/featureset.py [hunk 24]
```diff
@@ -0,0 +1,361 @@
...
+        graph = self.spec.graph
+        if state_name in graph.states:
+            state = graph.states[state_name]
+            aggregations = state.class_args.get("aggregates", [])
+            aggregations.append(aggregation)
+            state.class_args["aggregates"] = aggregations
+        else:
+            graph.add_step(
+                name=state_name,
+                after=after or "$prev",
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/model/featureset.py:303]: done

File: mlrun/featurestore/model/validators.py [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+
+    def check(self, value):
+        return True, {}
+
+
+class MinMaxValidator(Validator):
+    """validate min/max value ranges"""
+
+    kind = "minmax"
+    _dict_fields = ["kind", "check_type", "severity", "min", "max"]
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/model/validators.py:26]: done

File: mlrun/featurestore/model/validators.py [hunk 26]
```diff
@@ -0,0 +1,62 @@
...
+                        {
+                            "message": "value is greater than max",
+                            "max": self.max,
+                            "value": value,
+                        },
+                    )
+        return ok, args
+
+
+validator_types = {
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/model/validators.py:59]: done

File: mlrun/featurestore/sources.py [hunk 27]
```diff
@@ -0,0 +1,87 @@
...
+
+        return storey.DataframeSource(
+            dfs=self._df, key_column=key_column, time_column=time_column,
+        )
+
+
+source_kind_to_driver = {
+    "": BaseSourceDriver,
+    "csv": CSVSource,
+}
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/sources.py:87]: added comment

File: tests/feature-store/test_feature_store.py (hunk 1 above)
Comment (yaronha) [reply tests/feature-store/test_feature_store.py:23]: done

File: tests/feature-store/test_infer.py (hunk 2 above)
Comment (yaronha) [reply tests/feature-store/test_infer.py:28]: done

File: mlrun/featurestore/model/datatypes.py [hunk 28]
```diff
@@ -0,0 +1,73 @@
...
+        "integer": ValueType.INT64,
+        "string": ValueType.STRING,
+        "number": ValueType.DOUBLE,
+        "datetime": ValueType.DATETIME,
+        "boolean": ValueType.BOOL,
+    }
+    return type_map[value]
+
+
+def py_to_value_type(value):
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/model/datatypes.py:50]: done

File: mlrun/featurestore/model/base.py [hunk 29]
```diff
@@ -0,0 +1,270 @@
...
+
+    def __dir__(self):
+        return list(self._config) + dir(self.__class__)
+
+    def __repr__(self):
+        name = self.__class__.__name__
+        return f"{name}({self._config!r})"
+
+
+store_config = FeatureStoreConfig(copy.deepcopy(default_config))
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/model/base.py:74]: done, had to add couple of methods to `config`

File: mlrun/featurestore/model/vector.py [hunk 30]
```diff
@@ -0,0 +1,267 @@
...
+from ..targets import get_offline_target
+from ...datastore import get_store_uri
+from ...utils import StorePrefix
+
+
+class FeatureVectorError(Exception):
+    """ feature vector error. """
+
+    def __init__(self, *args, **kwargs):
+        pass
```
[the remark this answers is not in the document]
Comment (yaronha) [reply mlrun/featurestore/model/vector.py:34]: done

**Review by Hedingber — APPROVED** [review]

**Pull request merged**